### DISCUSSION TITLE
image pre-processing before model training

### DISCUSSION THREAD

**Discussion by akraskovSMDB** [neuropoly/axondeepseg#745] [Help needed]
Hi, 

in the fairly old discussion https://github.com/axondeepseg/axondeepseg/pull/50 @alzaia mentioned several image pre-processing steps before image segmentation. 
I could not find exact steps, if it is a normalisation what are the cut of limits, is histogram equalization done as well (also mentioned in https://github.com/ivadomed/ivadomed/issues/26)?
Are these steps performed on images to be segmented or also on the images which are used to train a new model?

Thank you

**Comment by mathieuboudreau** [comment]
Hi @akraskovSMDB !

That pull request was for a much older version of ADS. While there are some image pre-processing steps used for training in IVADOMED, for the segmentation step (which I believe is your concern, and what that PR handled), I believe the only preprocessing steps we do are 1) image resampling (if the image resolution is not the same as the resolution used in training, and/or if a zoom factor is used) and "patching" (cutting the image into the same size, eg 512x512, usually with some overlap), although this later step can be omitted in the option ("no-patch" flag in either the command line or Napari plugin options). There is nothing like histogram equalization done on the ADS part of the code, and I think the issue you linked in Napari relates to the MRI segmentation application side of their code and not necessarily (but maybe?) the histology-side.

Do you need to know the exact steps for a specific purpose (eg, reporting it in a publication)?

**Comment by akraskovSMDB** [comment]
It would be useful to know the precise steps, eg for publication. But also to know whether I need to do any pre-processing myself. 

I am interested in pre-processing not only for segmentation but also for model training.
In general wouldn't it be a good idea to have the same pre-processing steps for images used for model training and for further segmentation? Or U-NET  is so clever that it will work in the same way even if, e.g. contrasts, are different between images used for training and for segmentation?

**Reply by mathieuboudreau** [reply]
>Or U-NET is so clever that it will work in the same way even if, e.g. contrasts, are different between images used for training and for segmentation?

No, it should be the same contrast. We don't do this type of image preprocessing anymore (eg inverting contrast).

>In general wouldn't it be a good idea to have the same pre-processing steps for images used for model training and for further segmentation?

Yes, and they are mostly the same. IVADOMED does however do some data augmentation pre-processng steps for training though (i.e. rotating images, translating them, etc), in order to increase the number of "training" images. This also makes the model more robust I believe. But these aren't done for the segmentation step. For more details about the training preprocessing steps, I'd suggest you inquire with IVADOMED, or maybe @mariehbourget can chime in briefly if she recalls and has the time.

**Comment by akraskovSMDB** [comment]
> No, it should be the same contrast. We don't do this type of image preprocessing anymore (eg inverting contrast).

I meant something more like histogram stretching rather than inversion. Although when I was testing your TEM and SEM models on my images it took me some time to realise that you have opposite contrasts for TEM (axon back) and SEM (axon white). 

I played a bit with different pre-processing. My original data is RGB, 8 bits, Firstly what does the following line which appears three times per image in the segmentation log means 
"Lossy conversion from float32 to uint8. Range [43.6609992980957, 241.16299438476562]. Convert image to uint8 prior to saving to suppress this warning.
Lossy conversion from float32 to uint8. Range [43.6609992980957, 241.16299438476562]. Convert image to uint8 prior to saving to suppress this warning.
Lossy conversion from float32 to uint8. Range [43.6609992980957, 241.16299438476562]. Convert image to uint8 prior to saving to suppress this warning
"
The image is uint8 but just had three colour channels in it. Why is it being converted and to what? Are all RGB channels taken into account for segmentation? Or RGB is converted to some kind of intensity image?


Secondly, I  stretched histogram by clipping 0.1% of pixel intensities on both ends (first converted to YCbCr, stretched only Y channel and then converted back to RGB).  The same conversion line is still in the segmentation log file (x3), but the range is wider (as expected)  
Lossy conversion from float32 to uint8. Range [4.104000091552734, 253.86000061035156]. Convert image to uint8 prior to saving to suppress this warning.
Why, what does it mean?

Thirdly, I converted RGB into Gray (+ stretched), no conversion line in log file. I noticed that the number of axial patches is identical to RGB. Does it mean that each RGB patch is treated as a single patch with just three channels in it? Or there is some conversion to a single channel going on?

I also tried each colour channel separately (stretched in the same way).

I compared dice scores for each of those pre-processing steps. It turns out that Original RGB, stretched RGB and grey images give almost identical results in terms of dice score and visual inspection of the segmentations confirms this. Individual channels give worse results, especially the channel which has most of the background from the stain (blue from toluidine blue).
  
Sorry, after the long description of the method, the questions are 
What kind of pre-processing is happening during segmentation? 
Are RGB images converted into single-channel intensity images? (Almost identical results of segmentations of grey and RGB and identical number of patches for Grey and RGB (unless RGB patch is 3 dimensional) suggest this).

What does the warning "Lossy conversion from float32 to uint8. Range [4.104000091552734, 253.86000061035156]. Convert image to uint8 prior to saving to suppress this warning." mean?

Thank you and sorry for the long post

**Reply by mathieuboudreau** [reply]
>Are RGB images converted into single-channel intensity images? (Almost identical results of segmentations of grey and RGB and identical number of patches for Grey and RGB (unless RGB patch is 3 dimensional) suggest this).

Dealing with loading grayscale images vs rgb for the segmentation step is done on the IVADOMED side, and is handled here: https://github.com/ivadomed/ivadomed/blob/e101ebea632683d67deab3c50dd6b372207de2a9/ivadomed/loader/segmentation_pair.py#L318-L323

Basically, if you load an RGB image, we use imageio's "as_gray" option to bring it into a grayscale image.

As for the warnings, they appear because the images are not 8bit, and we convert them to 8bit. It's just a warning, and shouldn't be of a concern, it's just letting you know that you might be losing "information", but that's because you're condensing a high-bit (i.e. range of values) to a lower bit image.

In terms of preprocessing, I don't think we do any of the kinds you mentioned i.e histogram manipulation.


If you've got a lot more questions or would want to discuss a bit more in-depth, you could reach out via email and we could setup a Zoom call to chat!

**Reply by akraskovSMDB** [reply]
Dear @mathieuboudreau, thank you for your responses and for the offer of zoom call. 

Just to confirm, the same "treatment" of RGB images (essentially converting them into one-dimensional grey images) is happening during model training as well, correct? 
I also noticed the following comment "TODO: (#739) implement OMETIFF behavior (if "ome" in extension)". Does it mean that ivadomed is not able yet to use multidimensional data during training and segmentation? 

For the snipped of the code you've sent it seems that reduction (from 3d to 2d, ie from RGB to grey) is happening during conversion from png, tif to nifti. Is it possible to skip conversion from tiff, png into nifti and use instead nifti to start with? And if yes, can this nifti be multidimensional: imagessize_x x imagesize_y x modality (eg, different colour channels)? Will the third dimension be taken into account or ignored?

thank you

**Reply by mathieuboudreau** [reply]
>Just to confirm, the same "treatment" of RGB images (essentially converting them into one-dimensional grey images) is happening during model training as well, correct?

Models are trained on 8bit grayscaled images, yes!

>I also noticed the following comment "TODO: (https://github.com/axondeepseg/axondeepseg/pull/739) implement OMETIFF behavior (if "ome" in extension)". Does it mean that ivadomed is not able yet to use multidimensional data during training and segmentation?

I think you pointed to the wrong pull request - I don't see OMETIFF mentionned there. Nonetheless, I'm not aware of IVADOMED's capability on training on multdimensional data - that's a question that would be better asked to the IVADOMED project maintainers
.
>For the snipped of the code you've sent it seems that reduction (from 3d to 2d, ie from RGB to grey) is happening during conversion from png, tif to nifti. Is it possible to skip conversion from tiff, png into nifti and use instead nifti to start with? And if yes, can this nifti be multidimensional: imagessize_x x imagesize_y x modality (eg, different colour channels)? Will the third dimension be taken into account or ignored?

For the histology part of IVADOMED (what ADS uses), I don't think so. IVADOMED also trains on MRI  data, which may be loaded directly from NIFTI, so it's likely that this is possible, just not in the code that we use for ADS. Another question that could be asked there: https://github.com/ivadomed/ivadomed/issues
